pageserver/src/tenant/secondary.rs

```diff
@@ -95,7 +96,41 @@ pub(crate) struct SecondaryTenant {
     shard_identity: ShardIdentity,
     tenant_conf: std::sync::Mutex<TenantConfOpt>,
 
+    // Internal state used by the Downloader.
     detail: std::sync::Mutex<SecondaryDetail>,
+
+    // Public state indicating overall progress of downloads relative to the last heatmap seen
+    pub(crate) progress: std::sync::Mutex<SecondaryProgress>,
+}
+
+/// The progress of a secondary tenant is mostly useful when doing a long running download: e.g. initiating
+/// a download job, timing out while waiting for it to run, and then inspecting this status to understand
+/// what's happening.
+#[derive(Default, Debug, Serialize, Deserialize, Clone)]
+pub(crate) struct SecondaryProgress {
+    /// The remote storage LastModified time of the heatmap object we last downloaded.
+    #[serde(serialize_with = "ser_rfc3339_millis")]
+    heatmap_mtime: Option<SystemTime>,
+
+    /// The number of layers currently on-disk
+    layers_downloaded: usize,
+    /// The number of layers in the most recently seen heatmap
+    layers_total: usize,
+
+    /// The number of layers currently on-disk
+    bytes_downloaded: u64,
+    /// The number of layers in the most recently seen heatmap
+    bytes_total: u64,
+}
+
+fn ser_rfc3339_millis<S: serde::Serializer>(
```

Comment:
Yeah, I played with that: didn't find a neat way to re-use them -- since they're basically one-liners anyway, once you make a wrapper for the Option case, you've got basically the same lines as just writing it out like this.

Implementing Serialize for SystemTime so that we consistently encoding them all the same way would be a good thing -- probably beyond the scope of this PR though.

Reply:
Should we open a follow up and label it as `good-first-issue`. Maybe we get lucky and someone does it for us :)

Reply:
https://github.com/neondatabase/neon/issues/7151